site/e2e/reporter.ts

```diff
@@ -1,85 +1,136 @@
-import fs from "fs";
+/* eslint-disable no-console -- Logging is sort of the whole point here */
+import * as fs from "fs/promises";
 import type {
   FullConfig,
   Suite,
   TestCase,
   TestResult,
   FullResult,
   Reporter,
+  TestError,
 } from "@playwright/test/reporter";
 import axios from "axios";
+import type { Writable } from "stream";
 
 class CoderReporter implements Reporter {
+  config: FullConfig | null = null;
+  testOutput = new Map<string, Array<[Writable, string]>>();
+  passedCount = 0;
+  failedTests: TestCase[] = [];
+  timedOutTests: TestCase[] = [];
+
   onBegin(config: FullConfig, suite: Suite) {
-    // eslint-disable-next-line no-console -- Helpful for debugging
-    console.log(`Starting the run with ${suite.allTests().length} tests`);
+    this.config = config;
+    console.log(`==> Running ${suite.allTests().length} tests`);
   }
 
   onTestBegin(test: TestCase) {
-    // eslint-disable-next-line no-console -- Helpful for debugging
-    console.log(`Starting test ${test.title}`);
+    this.testOutput.set(test.id, []);
+    console.log(`==> Starting test ${test.title}`);
   }
 
-  onStdOut(chunk: string, test: TestCase, _: TestResult): void {
-    // eslint-disable-next-line no-console -- Helpful for debugging
-    console.log(
-      `[stdout] [${test ? test.title : "unknown"}]: ${chunk.replace(
-        /\n$/g,
-        "",
-      )}`,
-    );
+  onStdOut(chunk: string, test?: TestCase, _?: TestResult): void {
+    if (!test) {
+      console.log(`[stdout] ${chunk.replace(/\n$/g, "")}`);
+      return;
+    }
+    this.testOutput.get(test.id)!.push([process.stdout, chunk]);
```

Comment:
I know that ideally, the map should always have an array to give back, but what happens if there is a mismatch for the test ID, the push happens on an undefined value, and an error happens? Does Playwright have a way to recover from that?

I don't think there's anything wrong with the logic – I just naturally worry about things like this when I see non-null assertions

Reply:
I think that would be a pretty severe but in playwright, that we can deal with if it actually happens :p